include/cpp2util.h

```diff
@@ -1846,23 +2003,26 @@ CPP2_FORCE_INLINE constexpr auto cmp_greater_eq(auto&& t, auto&& u) -> decltype(
 //
 //-----------------------------------------------------------------------
 //
-template< typename C >
-inline constexpr auto as_( auto&& x ) -> decltype(auto)
-{
-    if constexpr (narrowing<C, CPP2_TYPEOF(x)>) {
+template <typename C, typename X > 
+auto as_( X&& x ) -> decltype(auto) 
+{ 
+    //std::cout << __PRETTY_FUNCTION__ << " " << __LINE__ << std::endl;
+    if constexpr ( requires { {as<C>(std::forward<X>(x))} -> std::same_as<nonesuch_<casting_errors::narrowing_cast>>; } ) {
         static_assert(
-            program_violates_type_safety_guarantee<C, decltype(x)>,
-            "'as' does not allow unsafe narrowing conversions - if you're sure you want this, use `unsafe_narrow<T>()` to force the conversion"
-        );
-    }
-    else if constexpr( requires{ { as<C>(CPP2_FORWARD(x)) } -> std::same_as<nonesuch_>; } ) {
+            program_violates_type_safety_guarantee<C, X>,
+            "'as' does not allow unsafe narrowing conversions - if you're sure you want this, use `unsafe_narrow<T>()` to force the conversion");
+    } else if constexpr ( requires { {as<C>(std::forward<X>(x))} -> std::same_as<nonesuch_<casting_errors::no_to_string_cast>>; } ) {
         static_assert(
-            program_violates_type_safety_guarantee<C, decltype(x)>,
-            "No safe 'as' cast available - please check your cast"
-        );
-    }
-    //  else
-    return as<C>(CPP2_FORWARD(x));
+            program_violates_type_safety_guarantee<C, X>,
+            "No cpp2::to_string overload exists for this type!");
```

Comment:
How is this violated?
There's a catch-all overload `inline auto to_string(...) -> std::string`.

Reply:
Yes, that is right. There is a catch-all overload. That works for `cpp2::to_string` but is incompatible with the behavior of `as` when the cast is impossible.

The contract for `as` is that when you call it, you either have the correct output or an error (`static_assert` when possible and an exception on runtime).

Having `as` silently return `"(customize me - no cpp2::to_string overload exists for this type)" breaks that contract. We can detect that on compile time, so I think `static_assert` is the right choice. 

A similar thing will happen when you try to cast `int` to `size_t` (`size_t` cannot represent all values of `int`, so you will get static_assert with a suggestion to use `unsafe_narrow<T>()`

Reply:
I agree that choosing that overload is wrong.
It's good to know that using `as std::string` doesn't choose it.
I guess it works this way because the concept that checks for the overload checks for conversion to function pointer, and the `(...)` parameter list doesn't work with those.

Reply:
Yes. It does not cast.

Reply:
That is why I introduce this concept.